### PR TITLE
pip: generate and install man pages

### PR DESCRIPTION
**Pull request by anund**
## Description of changes

Generates pip man pages following the arch pip package or the sketch of a working change. As this is a change to pip's definition the full build hasn't been tested on a machine I own.

1. Move scipy onto `pypaInstallHook` avoiding infinite recursion from `pipInstallHook` when scipy is pulled in via `checks` in either `myst-parser` or `sphinx`
2.  Add `sphinxcontrib-towncrier` based on the [arch package](https://gitlab.archlinux.org/archlinux/packaging/packages/python-sphinxcontrib-towncrier)
3. Invokes and `pip_sphinxext.py` to build and pass definitions to `sphinx`.

Couple things are some amount of incomplete

1. sphinx package references some? but no where near all sphinx packages
2. patch to `sphinxcontrib-towncrier` is blindly copied
3. `intersphinx` is part of the pip docs and wants to pull references from the internet. `pythonDocs` appears to have the needed python object file reference but `pypug` is not? currently packaged.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/pip/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/pip/default.nix:69]: Have you using `sphinxHook` instead? Just out of curiosity.

File: pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix:18]: Please comment about the source of the patch, whether it should me removed on the next version bump of the file, and if there is an attempt to contribute the patch upstream. Also, I personally prefer this formatting:

```nix
patches = [
  # Comment here
  ./file.patch
];
```

**Review by doronbehar — CHANGES_REQUESTED** [review]
1. The changes IMO should be split to separate commits:

```
python3.pkgs.scipy: use `pypaInstallHook` instead of `pipInstallHook`
python3.pkgs.sphinx-contrib-towncrier: init at 0.3.2a0
python3.pkgs.pip: install man pages
```

2. Why is this a draft? Changes look good overall.

**Comment by doronbehar** [comment]
> 2\. Why is this a draft? Changes look good overall.

Oh now I read your top comment.

File: pkgs/development/python-modules/pip/default.nix
Comment (anund) [reply pkgs/development/python-modules/pip/default.nix:69]: Initially I looked at it as incompatible because of the pre-req coming from `python pip_sphinxext.py`. I'll have a look at using the hook just for this step and let you know.

**Comment by anund** [comment]
Any suggestion about when or if this should target `staging` and how to validate if it's broken something unintentional? Pip being the bottom of ~all python packages means I can't really fully test this locally. (even though it "looks" innocuous)

**Comment by anund** [comment]
Fished around for a commit that didn't involve building gcc and found this: https://hydra.nixos.org/jobset/nixpkgs/python-updates/evals. I'll use cdeab96 while staging hasn't been built.

**Comment by FRidh** [comment]
I split the scipy part off in https://github.com/NixOS/nixpkgs/pull/250509

**Comment by FRidh** [comment]
> Pip being the bottom of ~all python packages means I can't really fully test this locally. (even though it "looks" innocuous). I'm unfamiliar with how PR's of this kind might need special handling.

Not anymore. With the changes on staging pip isn't a python bootstrapping package anymore. It is however still used by some packages at times.

**Comment by anund** [comment]
@FRidh it looks like pip goes from ~40000 -> 3089 dependants. 1924 are `python3.1{0,1}*` and 739 are homeassistant-test. I have a copy of the nixpkgs-review output set if needed. That took ~8 hours of build time so I won't repeat generating it locally.

@doronbehar 
1. Regarding Sphinx package having a few contrib packages. They were added ~early on and the pattern doesn't appear to have been continued
2. Regarding the patch: seems to be arch build ecosystem specific, removed as unneeded.
3. `intersphinx` is needed for html generation, snipped out with sed + explanation comment.
4. Regarding `sphinxHook`: technically I can make it work with a few extra steps. Pip's sphinx setup is a bit non-standard? so I figure matching it is less likely to surprise in future changing files to line up with the current expectations of `sphinxHook` (namely `conf.py` is in an admittedly weird place in pip source)

File: pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix:18]: Nit:

```suggestion
  propagatedBuildInputs = [
    sphinx
    towncrier
    tomli
  ];
```

File: pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix:25]: No maintainers?

File: pkgs/development/python-modules/pip/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/pip/default.nix:57]: Left over from when testing `sphinxHook`?

File: pkgs/development/python-modules/pip/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/pip/default.nix:64]: Perhaps I'd comment here that using `sphinxHook` is not easier then running these commands yourself, due to pip putting docs in non expected locations.

File: pkgs/development/python-modules/pip/default.nix
Comment (doronbehar) [inline pkgs/development/python-modules/pip/default.nix:67]: Reading [the file](https://github.com/pypa/pip/blob/23.2.1/docs/html/conf.py) now, it should be safe, or even safer to use `substituteInPlace`. This way you'd at least get a warning in the future when the substitution doesn't apply anymore.

File: pkgs/development/python-modules/pip/default.nix
Comment (anund) [reply pkgs/development/python-modules/pip/default.nix:57]: Indeed.

File: pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix
Comment (anund) [reply pkgs/development/python-modules/sphinxcontrib-towncrier/default.nix]: Dropping. Looking around it's more maintained than unmaintained sphinxcontrib modules. I won't make it worse.

File: pkgs/development/python-modules/pip/default.nix
Comment (doronbehar) [reply pkgs/development/python-modules/pip/default.nix:64]: Maybe also the later argument could be an empty string - to simplify reading.

**Review by doronbehar — COMMENTED** [review]
Well done for removing the unnecessary sphinx plugins!

**Comment by doronbehar** [comment]
Seems pretty good to me overall.

File: pkgs/development/python-modules/pip/default.nix
Comment (anund) [reply pkgs/development/python-modules/pip/default.nix:64]: I'll switch to empty string for reading. A for loop only really makes 4 lines into 3 so I'd prefer the straight forward copy + paste at this time.

**Comment by doronbehar** [comment]
Seems perfect to me now. Let's wait a few more days for this to get more comments/  hopefully a final word from FRidh as well.

**Review by doronbehar — APPROVED** [review]
Built pip locally, the build log looks good, as well as the manpages themselves.

**Pull request merged**
